CI: Run gtest with BullsEye

Alexey-Rivkin opened the pull request:
## What?
Integrate Bullseye instrumentation into gtest.

## Why?
Enable collection, aggregation, and reporting of code coverage data.
[JIRA: HPCINFRA-1921](https://jirasw.nvidia.com/browse/HPCINFRA-1921) (Internal link).

## How?
1. Introduce a `run_bullseye` test parameter for granular control per specific host groups.
3. When enabled, Bullseye instrumentation is applied to gtests build and run process, collecting test coverage data.
5. Coverage reports from individual jobs are merged into a single report file and stored as a pipeline artifact.

### Exclusions
While this PR enables Bullseye for gtests, the following exclusions apply:
- GPU tests are excluded due to GPU driver failures when running with Bullseye.
- BlueField nodes (ARM architecture) are excluded due to instrumentation issues causing linker errors.
- RoCe tests with the protocol disabled, as we cover the same tests with the protocol enabled already.

Alexey-Rivkin committed:
CI: Run gtest with BullsEye

Alexey-Rivkin committed:
CI: Disable BullsEye for GPU tests.

Running Gtest with BullsEye caused GPU driver failures:
  NVRM: nvAssertFailedNoLog: Assertion failed: pIOVAS != NULL @ io_vaspace.c:591

Alexey-Rivkin committed:
CI: Disable BullsEye for BlueField nodes

BullsEye on ARM arch fails to properly instrument some code sections, resulting in linker errors:
"undefined reference to `cov_probe_v12'"

dpressle commented on contrib/test_jenkins.sh:
```diff
@@ -1070,6 +1070,41 @@ run_gtest_armclang() {
+
+	if ! command -v cov01 &> /dev/null; then
+		azure_log_warning "=== Skipping Bullseye: cov01 not found ==="
+		return 0
```
this means the test didnt run at all but you marked the stage as successful

Alexey-Rivkin replied:
[CI: Ensure Gtest always runs, regardless of Bullseye status](https://github.com/openucx/ucx/pull/10565/commits/8e7a45fd549501cba35ba7d32cb92878d36e364a)

dpressle commented on contrib/test_jenkins.sh:
```diff
@@ -1070,6 +1070,41 @@ run_gtest_armclang() {
+		*)       SRC_PATH="/auto/app/BullsEye/BullseyeCoverage-Latest-x86" ;;
+	esac
+
+	# Create a local copy of Bullseye binaries
```
try to explain why and not only what

Alexey-Rivkin committed:
CI: Ensure Gtest always runs, regardless of Bullseye status

dpressle commented on contrib/test_jenkins.sh:
```diff
@@ -1070,6 +1070,38 @@ run_gtest_armclang() {
+
+	build devel --enable-gtest
+	run_gtest "default"
+	cov01 --off
```
this will fail in case we dont have BC installed and we run tests without it

Alexey-Rivkin replied:
[CI: Run Bullseye post-processing only when enabled](https://github.com/openucx/ucx/pull/10565/commits/de11346d28c0689668716420ac7622c35314863f)

Alexey-Rivkin committed:
CI: Run Bullseye post-processing only when enabled

Alexey-Rivkin committed:
CI: Fix Indentation

Alexey-Rivkin commented on contrib/test_jenkins.sh:
```diff
@@ -1070,6 +1070,44 @@ run_gtest_armclang() {
+        if ! covfn -f "${COVFILE}"; then
+            azure_log_error "Bullseye report validation failed!"
+        fi
+    fi
```
[CI: Fix Indentation](https://github.com/openucx/ucx/pull/10565/commits/bc0a7678c762bae1a3941d1c0f694b35ee3e8ba2)

dpressle commented on buildlib/pr/main.yml:
```diff
@@ -322,34 +322,68 @@ stages:
+        name: MLNX
+        demands: ucx_docker
+      steps:
+      - bash: |
```
what will happen if we ended up not sunning with bullseye ? maybe we need to check if cov exists like you do in the test

Alexey-Rivkin replied:
The dedicated Bullseye job should fail in this case to avoid hiding issues

dpressle commented:
check why its not running on roce_proto_disable i dont see any messages about it
https://dev.azure.com/ucfconsort/ucx/_build/results?buildId=100525&view=logs&j=673b6020-2d94-5761-4112-cf6dfb119464&t=617e344f-4a4d-5f1b-5546-900cab530d2c

Alexey-Rivkin committed:
CI: Disable BullsEye for RoCE with protocol disabled

Bullseye is already being run for RoCE tests with the protocol enabled, which covers the majority of relevant code paths for these tests. Running Bullseye again with the protocol disabled provides minimal additional coverage, as protocol-disabled scenarios primarily skip or bypass certain code paths rather than introducing significant new behavior

Alexey-Rivkin commented:
> check why its not running on roce_proto_disable i dont see any messages about it https://dev.azure.com/ucfconsort/ucx/_build/results?buildId=100525&view=logs&j=673b6020-2d94-5761-4112-cf6dfb119464&t=617e344f-4a4d-5f1b-5546-900cab530d2c

Good catch!
[CI: Disable BullsEye for RoCE with protocol disabled](https://github.com/openucx/ucx/pull/10565/commits/0a6cc1c76b270cdc921b74544ef78ffa857917dc)

Alexey-Rivkin commented on contrib/test_jenkins.sh:
```diff
@@ -1070,6 +1070,44 @@ run_gtest_armclang() {
+	else
+		echo "=== Enable Bullseye instrumentation ==="
+		BULLSEYE_ENABLED=true
+		COV_DIR=/hpc/scrap/azure/bullseye/${BUILD_BUILDID}-${BUILD_BUILDNUMBER}
```
`/hpc/scrap/azure` is a dedicated shared workspace. 
We use the same approach in the performance testing pipeline.

Alexey-Rivkin replied:
Local dirs won’t work here since we need to aggregate report files from multiple machines for merging.

yosefe replied:
can we publish them as artifacts and use in next stage?

Alexey-Rivkin commented on contrib/test_jenkins.sh:
```diff
@@ -1070,6 +1070,44 @@ run_gtest_armclang() {
+	# Create a local Bullseye copy due to NFS instability
+	DEST_PATH="${WORKSPACE}/BS_LOCAL"
+	mkdir -p "$DEST_PATH"
+	rsync -az "$SRC_PATH/" "$DEST_PATH/" || true
```
`cp` doesn't preserve symlinks, which are essential for Bullseye's library structure.  
Failure is ignored to allow gtest to run even if Bullseye isn't present.

yosefe replied:
i think "cp -a" preserves symlinks
and we could check if BE dir presetn and exit (with warning) in the beginning of this function

Alexey-Rivkin replied:
We already check for `cov01` command availability later on. 
Would checking the directory bring added value or redundancy?

Alexey-Rivkin committed:
CI: Implement remarks

Alexey-Rivkin commented on contrib/test_jenkins.sh:
```diff
@@ -1070,6 +1070,44 @@ run_gtest_armclang() {
+	esac
+
+	# Create a local Bullseye copy due to NFS instability
+	DEST_PATH="${WORKSPACE}/BS_LOCAL"
```
[CI: Implement remarks](https://github.com/openucx/ucx/pull/10565/commits/1de8a9618fdf31693a6d2502eb060f71fa728454)

Alexey-Rivkin commented on contrib/test_jenkins.sh:
```diff
@@ -1070,6 +1070,44 @@ run_gtest_armclang() {
+
+	export PATH="$DEST_PATH/bin:$PATH"
+
+	if ! command -v cov01 &> /dev/null; then
```
1. [CI: Implement remarks](https://github.com/openucx/ucx/pull/10565/commits/1de8a9618fdf31693a6d2502eb060f71fa728454)
2. Just to confirm: are we supposed to fail tests if Bullseye isn’t available? Previously, we treated Bullseye as best-effort.

Alexey-Rivkin committed:
CI: Update Docker mounts and Bullseye coverage workflow

- Remove unnecessary volume mounts (/auto/app and /hpc/scrap) that were
  only needed for the previous Bullseye version
- Refactor the coverage merging process with new Bullseye location
- Add HTML report generation for better visualization of coverage results
- Implement path normalization for more readable reports
- Update workflow to download coverage files instead of direct file access

These changes align with the Bullseye upgrade to version 9.14.2 and
improve the coverage reporting process by centralizing artifacts in
the build pipeline rather than using external storage locations.

Alexey-Rivkin committed:
CI: Update Bullseye path and improve coverage collection

- Update Bullseye path to use version 9.14.2 from /auto/sw_tools
- Store coverage files in artifact staging directory instead of /hpc/scrap
- Exclude test directory from coverage with covselect
- Normalize workdir paths in coverage files for better report merging
- Replace rsync with cp for more reliable local copy creation
- Improve error handling and validation

These changes align with the upgraded Bullseye tooling and improve the
reliability of the coverage reporting process.

Alexey-Rivkin committed:
CI: Add partial coverage file publishing to pipeline

Add a step to publish partial coverage files from test runs when Bullseye
is enabled. This allows the coverage merge job to have access to all the
individual coverage reports generated during testing.

dpressle commented on contrib/test_jenkins.sh:
```diff
@@ -1070,6 +1076,43 @@ run_gtest_armclang() {
 
+run_gtest_bullseye() {
+	case "$AGENT_OSARCHITECTURE" in
+		"ARM64") bullseye="/auto/sw_tools/Bullseye/linux/linux64/9.14.2/aarch64/" ;;
```
remove last / and add it in the export lines below its not clear to see ${bullseye}lib it is better to see ${bullseye}/lib

Alexey-Rivkin replied:
Right, it hurt my eyes too

dpressle commented on contrib/test_jenkins.sh:
```diff
@@ -1070,6 +1076,43 @@ run_gtest_armclang() {
+	export LD_LIBRARY_PATH=${bullseye}lib:$LD_LIBRARY_PATH
+	export CPATH=${bullseye}include:$CPATH
+	export LD_RUN_PATH="${bullseye}lib:$LD_RUN_PATH"
+	export C_INCLUDE_PATH=${bullseye}include:$C_INCLUDE_PATH
```
shouldn't most of this exports only be done inside the if at line 1093? its useless unless running with BE

Alexey-Rivkin replied:
Most of those export statements are likely unnecessary and were just added for troubleshooting

Alexey-Rivkin committed:
CI: Add Bullseye coverage exclusion configuration

Creates a dedicated configuration file for excluding specific
directories and test functions from Bullseye coverage reports.

Alexey-Rivkin committed:
CI: Switch Bullseye coverage from RoCE to GPU

The new Bullseye version works well with GPU tests, but fails on RoCE tests compilation.

Alexey-Rivkin committed:
CI: Enhance Bullseye integration reliability

Adds instrumentation verification, sets proper environment variables,
and uses centralized exclusion configuration. Makes coverage collection
more reliable across different tests.

dpressle commented on contrib/test_jenkins.sh:
```diff
@@ -1070,6 +1070,43 @@ run_gtest_armclang() {
+		*)       bullseye="/auto/sw_tools/Bullseye/linux/linux64/9.14.2/amd64" ;;
+	esac
+
+	export PATH="${bullseye}/bin:$PATH"
```
I dont think we want this in case of non BE so why wont we add this in the cov01 if and export it only in case we do have BE
if PATH="${bullseye}/bin:$PATH" command -v cov01 ;

Alexey-Rivkin replied:
Switched to env. modules instead

dpressle reviewed (changes requested):
for GPU need to disable perf - TEST_PERF=0

Alexey-Rivkin committed:
CI: Switch to Bullseye module

Alexey-Rivkin committed:
CI: Disable perf test with Bullseye

dpressle commented:
@yosefe  please review

Alexey-Rivkin commented:
@gleon99 still relevant?
